### ISSUE TITLE
v0.12.0 regression: Partial blocks are never deleted; spam in logs for double deletion mark.

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#2459]
It was hard to track down, but looks like 4 unexpected things are happening on v0.12.0:

1. Partial blocks (no meta.json) are never cleaned up: 
![image](https://user-images.githubusercontent.com/6950331/79596563-b1a6a080-80d8-11ea-84c7-05849e10ae4b.png)

2. We have the spam of double mark deletion for partial upload delete attempt (for each partial block):

```
l=info ts=2020-04-17T16:50:14.305651352Z caller=clean.go:52 msg="deleted aborted partial upload" block=01E5Q7MP483QSDSX036X4TZ5SD thresholdAge=48h0m0s
level=info ts=2020-04-17T16:50:14.305659796Z caller=clean.go:47 msg="found partially uploaded block; marking for deletion" block=01E50RCWDZ5P83NRH4R0GXVGCC
level=warn ts=2020-04-17T16:50:14.312057891Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E50RCWDZ5P83NRH4R0GXVGCC/deletion-mark.json already exists in bucket"
level=info ts=2020-04-17T16:50:14.312073215Z caller=clean.go:52 msg="deleted aborted partial upload" block=01E50RCWDZ5P83NRH4R0GXVGCC thresholdAge=48h0m0s
level=info ts=2020-04-17T16:50:14.312081259Z caller=clean.go:47 msg="found partially uploaded block; marking for deletion" block=01E5695C03VK6FB89G49T5MRB4
level=warn ts=2020-04-17T16:50:14.317892887Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E5695C03VK6FB89G49T5MRB4/deletion-mark.json already exists in bucket"
level=info ts=2020-04-17T16:50:14.317907097Z caller=clean.go:52 msg="deleted aborted partial upload" block=01E5695C03VK6FB89G49T5MRB4 thresholdAge=48h0m0s
level=info ts=2020-04-17T16:50:14.317914967Z caller=clean.go:47 msg="found partially uploaded block; marking for deletion" block=01E5J2V84SKW9GYMH8B88M5NR9
```

3. We have spam of double mark deletions for garbage collection attempts (thousands+ as well):

```
level=warn ts=2020-04-16T23:01:18.533895718Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E57S72A216JY3H6JXXK1GDE5/deletion-mark.json already exists in bucket"
level=info ts=2020-04-16T23:01:18.533922849Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E5F2TGZ6FD6KS5WDMKBKGKP5
level=warn ts=2020-04-16T23:01:18.540796799Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E5F2TGZ6FD6KS5WDMKBKGKP5/deletion-mark.json already exists in bucket"
level=info ts=2020-04-16T23:01:18.540817511Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E58037KF7XF4TFMDQZYFKJCT
level=warn ts=2020-04-16T23:01:18.546292591Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E58037KF7XF4TFMDQZYFKJCT/deletion-mark.json already exists in bucket"
level=info ts=2020-04-16T23:01:18.546319501Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E58037KFCFY22BZZR4RJKG7G
```

4. (theoretically) Malformed deletion mark can cause the problem: log and inability to delete it.

We found the root cause of 1, 2 and 4 still looking for 3, help wanted (: 3 is not very critical other than spam. 

Root cause explained on [diagram](https://docs.google.com/drawings/d/14hDdVsRg9k9CgbxtzMNnHHrvNfxlGbqwuiw5nNAmcFM/edit):
![image](https://user-images.githubusercontent.com/6950331/79597067-8a040800-80d9-11ea-8053-e376c887df11.png)

Fix in progress. Let's make sure it is in 0.12.1

cc @kakkoyun @squat @pstibrany @pracucci

**Comment by tik-stbuehler** [comment]
3 is also critical because it fills the disk - at least the log partition (if you have a separate one), otherwise the root filesystem. Any workaround? Is downgrade possible/an option?

**Comment by bwplotka** [comment]
Did you really experience this or you say theoretically?

We fixed the root cause of 1 and 2 and potentially 3 in v0.12.1

**Comment by tik-stbuehler** [comment]
Yes, I did experience this (about 15G logs on a 20G rootfs; compact has it's own partition) with v0.12.0.

While https://github.com/thanos-io/thanos/commits/release-0.12 seems to indicate that v0.12.1 has been released, there is no tag and no binaries (and no relase on https://github.com/thanos-io/thanos/releases). Any idea when those will surface?

**Comment by bwplotka** [comment]
We are in progress in cutting it. Once done we will let you know so you can
test with us if all works (:

**Comment by ThoreKr** [comment]
![image](https://user-images.githubusercontent.com/5465291/79863916-da7cad80-83d8-11ea-9ebb-fd9559b22668.png)

I'm not sure how this affects other s3 providers (and thereby cost), but on the radosgw we observed a lot of additional load from 3.

**Comment by bwplotka** [comment]
Yes that can be bad outcome indeed.

Ok, let's see if v0.12.1 helps, if not let's reopen and investigate this
issue.

**Comment by tik-stbuehler** [comment]
It looks better now; but I got a series of blocks (level 3, duration 2d) with the same resolution as a big block (level 4, duration 14d) - how can I clean that up (I think there are smaller overlaps too)?

```
|            ULID            |        FROM         |        UNTIL        |  RANGE   | UNTIL-DOWN |  #SERIES  |    #SAMPLES    |   #CHUNKS   | COMP-LEVEL | COMP-FAILED |              LABELS              | RESOLUTION |  SOURCE   |
|----------------------------|---------------------|---------------------|----------|------------|-----------|----------------|-------------|------------|-------------|----------------------------------|------------|-----------|
| 01E6AVP6GKWE3A1310J36DK26Q | 02-04-2020 10:00:00 | 16-04-2020 02:00:00 | 328h0m0s | -88h0m0s   | 1,359,236 | 4,237,094,273  | 37,028,353  | 4          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
| 01E51V6Q78879SSNAD7V1V9JFA | 02-04-2020 10:00:00 | 04-04-2020 02:00:00 | 40h0m0s  | 200h0m0s   | 1,132,551 | 514,393,151    | 4,362,598   | 3          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
| 01E56SHZBK95B34JEB3Q3WHWBS | 04-04-2020 02:00:00 | 06-04-2020 02:00:00 | 48h0m0s  | 192h0m0s   | 1,120,570 | 619,721,196    | 5,422,677   | 3          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
| 01E5BXX1BKQCDYG6TD64YT3ZNR | 06-04-2020 02:00:00 | 08-04-2020 02:00:00 | 48h0m0s  | 192h0m0s   | 1,147,611 | 612,630,958    | 5,450,633   | 3          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
| 01E5H2PCG7YBWKJ3RF53PS20X2 | 08-04-2020 02:00:00 | 10-04-2020 02:00:00 | 48h0m0s  | 192h0m0s   | 1,125,117 | 619,312,008    | 5,425,543   | 3          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
| 01E5P7GBXZDK0HMZY6VDGF5VXN | 10-04-2020 02:00:00 | 12-04-2020 02:00:00 | 48h0m0s  | 192h0m0s   | 1,103,131 | 616,139,172    | 5,380,980   | 3          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
| 01E5VC8MAR3MWWHG532PEE6Y34 | 12-04-2020 02:00:00 | 14-04-2020 02:00:00 | 48h0m0s  | 192h0m0s   | 1,105,801 | 617,179,789    | 5,389,887   | 3          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
| 01E60H2V5G1BP4VJ545B4362RV | 14-04-2020 02:00:00 | 16-04-2020 02:00:00 | 48h0m0s  | 192h0m0s   | 1,168,501 | 637,717,999    | 5,596,035   | 3          | false       | cluster=nks-prometheus,replica=A | 5m0s       | compactor |
```

**Comment by bwplotka** [comment]
You mean partial one?

**Comment by tik-stbuehler** [comment]
Well, this morning the smaller blocks have vanished, so it seems they got cleaned up after all, just with some delay.

**Comment by ThoreKr** [comment]
After 2 days in wait mode the compactor started to log-spam these again:

```
Apr 23 19:01:17 thanos-compactor thanos[319]: level=warn ts=2020-04-23T17:01:17.270548661Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E65TGR4ZET7WJAYA0J2SBAT9/deletion-mark.json already exists in bucket"
```

And with it again increased requests to s3.

This is similar to the issue in 0.12.0, where it also took a couple of days for this to kick in (runing the compactor with `--wait`)

**Comment by bwplotka** [comment]
Can you show full logs? And metrics for `thanos_meta_synced`? (:

Looks like item number `3` is still happening.

**Comment by ThoreKr** [comment]
```
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="corrupted-meta-json"}	0
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="duplicate"}	24
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="failed"}	0
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="label-excluded"}	0
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="loaded"}	187
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="marked-for-deletion"}	44
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="no-meta-json"}	92
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="time-excluded"}	0
thanos_blocks_meta_synced{instance="thanos-compactor:10902",job="thanos",state="too-fresh"}	0
```

It seems to generate these two lines

```
Apr 23 19:01:28 thanos-compactor thanos[319]: level=info ts=2020-04-23T17:01:28.374061053Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E698CC2G3RK05VTJ39ND2AVX
Apr 23 19:01:28 thanos-compactor thanos[319]: level=warn 
```
For 92 blocks

As I started the compactor again a couple of minutes ago, 184 of these log messages came up:

```
Apr 23 22:04:50 thanos-compactor thanos[11883]: level=info ts=2020-04-23T20:04:50.964227949Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E698CC2G3RK05VTJ39ND2AVX
Apr 23 22:04:50 thanos-compactor thanos[11883]: level=info ts=2020-04-23T20:04:50.983918997Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E698CC2G3RK05VTJ39ND2AVX thresholdAge=48h0m0s
```
I took a couple of random samples and I found the block IDs in both logs.
So I assume they are identical.

Unfortunately I do not have a backup of such a block, I assume in ~2 days it would be possible to download such a block.

**Comment by ThoreKr** [comment]
So, since restarting on the 23rd at around 10pm, today at around 5:30 am the load on the rgw increased again (~40 get ops/s, far from the roughly 400 on thursday).

However it was looping again over a few blocks, but they got cleaned up while I tried to download one of them.
Interestingly the number of requests has not decreased since then, though logs remain silent.

This is what happened today (antichronological, because loki..., the "investigate" spam reduced to one sample, but it looped all the time).
```
2020-04-25 12:33:56	level=info ts=2020-04-25T10:33:56.028492513Z caller=fetcher.go:451 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=403.620117ms cached=253 returned=253 partial=0
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.855909169Z caller=blocks_cleaner.go:57 msg="cleaning of blocks marked for deletion done"
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.855848574Z caller=blocks_cleaner.go:43 msg="started cleaning of blocks marked for deletion"
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.855808978Z caller=clean.go:60 msg="cleaning of aborted partial uploads done"
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.855760165Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6J1X5CZT7WFTVMD7P7DD2PS thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.837906511Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6J1X5CZT7WFTVMD7P7DD2PS
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.837854574Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6J8RWM4BNR585FTJDH659CG thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.821952478Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6J8RWM4BNR585FTJDH659CG
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.821899961Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6J8RWN01QTQ8JFDDZ67J4YZ thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.806310608Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6J8RWN01QTQ8JFDDZ67J4YZ
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.806268919Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6HM5PWVPPDN0SKZC7WJ6TWN thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.791343619Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6HM5PWVPPDN0SKZC7WJ6TWN
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.791285731Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6HM5PWZ71EAWMXAYGM3XT9K thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.775099953Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6HM5PWZ71EAWMXAYGM3XT9K
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.775042048Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6HV1E51HRYR4HHEDZZ709W8 thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.759119448Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6HV1E51HRYR4HHEDZZ709W8
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.75908082Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6J1X5CWE1616WRE9SBJGD2G thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.741433579Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6J1X5CWE1616WRE9SBJGD2G
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.741379614Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6HV1E4Z5XP1NE8APWP5K55R thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.720200087Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6HV1E4Z5XP1NE8APWP5K55R
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.720181507Z caller=clean.go:33 msg="started cleaning of aborted partial uploads"
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.720142206Z caller=retention.go:45 msg="optional retention apply done"
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.720041762Z caller=retention.go:30 msg="start optional retention"
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.708253413Z caller=fetcher.go:451 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=1.661918639s cached=253 returned=193 partial=8
2020-04-25 12:33:13	level=info ts=2020-04-25T10:33:13.046255042Z caller=compact.go:408 msg="downsampling iterations done"
2020-04-25 12:33:13	level=info ts=2020-04-25T10:33:13.000232064Z caller=fetcher.go:451 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=1.396778406s cached=253 returned=193 partial=8
2020-04-25 12:33:11	level=info ts=2020-04-25T10:33:11.603391711Z caller=compact.go:401 msg="start second pass of downsampling"
2020-04-25 12:33:11	level=info ts=2020-04-25T10:33:11.53074912Z caller=fetcher.go:451 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=2.483776175s cached=253 returned=193 partial=8
2020-04-25 12:33:09	level=info ts=2020-04-25T10:33:09.046744693Z caller=compact.go:393 msg="start first pass of downsampling"
2020-04-25 12:33:09	level=info ts=2020-04-25T10:33:09.03937177Z caller=compact.go:936 msg="compaction iterations done"
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.056848659Z caller=compact.go:904 msg="start of compactions"
2020-04-25 12:33:06	level=warn ts=2020-04-25T10:33:06.056201205Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E6J8RWM4BNR585FTJDH659CG/deletion-mark.json already exists in bucket"
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.054161951Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E6J8RWM4BNR585FTJDH659CG
2020-04-25 12:33:06	level=warn ts=2020-04-25T10:33:06.054120081Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E6J1X5CZT7WFTVMD7P7DD2PS/deletion-mark.json already exists in bucket"
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.051849649Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E6J1X5CZT7WFTVMD7P7DD2PS
2020-04-25 12:33:06	level=warn ts=2020-04-25T10:33:06.051803696Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E6HV1E4Z5XP1NE8APWP5K55R/deletion-mark.json already exists in bucket"
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.049216505Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E6HV1E4Z5XP1NE8APWP5K55R
2020-04-25 12:33:06	level=warn ts=2020-04-25T10:33:06.049171352Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E6HM5PWZ71EAWMXAYGM3XT9K/deletion-mark.json already exists in bucket"
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.046633978Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E6HM5PWZ71EAWMXAYGM3XT9K
2020-04-25 12:33:06	level=warn ts=2020-04-25T10:33:06.046590568Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E6J8RWM4BNR585FTJDH659CG/deletion-mark.json already exists in bucket"
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.044469206Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E6J8RWM4BNR585FTJDH659CG
```

**Comment by ThoreKr** [comment]
This patterns seems to repeat a for couple of seconds before deleting the partial uploads. I haven't managed yet to download one of those blocks.
Also on the rgw (s3 gateway) side the requests seem to pile up, it started with an average of around 10 requests, then increased to 40, now it is around 100).
In the log I see a lot of `HEAD` requests to the `meta.json` and `deletion-mark.json`.

Are statistics about this of interest?

**Comment by bwplotka** [comment]
Your points are super useful! This narrows down a problem a lot, we will dive to it again from tomorrow :hugs: 

Just curious, maybe something silly, but what if you restart compactor. Does this "pile" stays or get reset? (:

**Comment by ThoreKr** [comment]
It goes back to 0, let's see when this starts again. The behaviour is interesting and different to 12.0.0.

![image](https://user-images.githubusercontent.com/5465291/80317103-1ad99280-8802-11ea-94e7-daa0dc1cd93d.png)

Every 8h it seemed to stack up and then apparently reached some sort of limit. In 12.0.0 it it did get to the high number of requests almost immediately.

In parallel for each of these stages shows up in `thanos_blocks_meta_synced{state="no-meta-json"}`starting with 8, then 16 then 24, then 44, before restarting the compactor it got up to 52.
![image](https://user-images.githubusercontent.com/5465291/80317297-4ad56580-8803-11ea-9431-c75029d57822.png)


These blocks, seem to be just empty folders with a `deletion-mark.json`
```
{"id":"01E6HV1E51HRYR4HHEDZZ709W8","deletion_time":1587842576,"version":1}
```

**Comment by gburek-fastly** [comment]
We are seeing, when running 0.12.1, an increasing number of blocks with only the deletion-mark.json file.

```
thanos_blocks_meta_synced{job="thanos_store",state="corrupted-meta-json"} | 0
thanos_blocks_meta_synced{,job="thanos_store",state="failed"} | 0
thanos_blocks_meta_synced{job="thanos_store",state="label-excluded"} | 0
thanos_blocks_meta_synced{,job="thanos_store",state="loaded"} | 10603
thanos_blocks_meta_synced{job="thanos_store",state="no-meta-json"} | 11024
thanos_blocks_meta_synced{job="thanos_store",state="time-excluded"} | 2640
thanos_blocks_meta_synced{,job="thanos_store",state="too-fresh"} | 0
thanos_blocks_meta_synced{job="thanos_store",state="duplicate"} | 1914
thanos_blocks_meta_synced{job="thanos_store",state="marked-for-deletion"} | 0
```

Many of these blocks appear to be correctly deleted but the mark remains. It is unclear how many of these were from before the 0.12.0->0.12.1 upgrade or a config change where we went from default 48h to 6h delete-delay. 

This is weighing on our store nodes, where scanning these blocks, even when empty, has translated into process start up times of > 25min. 

![image](https://user-images.githubusercontent.com/49920412/80533204-1c19d500-8952-11ea-866b-8667576d25cc.png)

**Comment by bwplotka** [comment]
> This is weighing on our store nodes, where scanning these blocks, even
when empty, has translated into process start up times of > 25min.

25m of what exactly?

**Comment by gburek-fastly** [comment]
The init_duration as reported by store is what we have been using to track process strart up times. From our logs, we started with 8m as a base line before upgrading to 0.12.0

```
Apr 20 18:52:57 thanos[25225]: level=info ts=2020-04-20T18:52:57.35383862Z caller=store.go:305 msg="bucket store ready" init_duration=8m30.98069259s
```

After this, each init_duration became slower, as compact started using the mark-delete method. and our duplicate, marked-for-deletion and no-meta-json block count crept up. 
```
Apr 20 19:31:32 thanos[30956]: level=info ts=2020-04-20T19:31:32.351569057Z caller=store.go:305 msg="bucket store ready" init_duration=13m18.228384754s
Apr 20 23:34:02 thanos[30130]: level=info ts=2020-04-20T23:34:02.177312023Z caller=store.go:305 msg="bucket store ready" init_duration=14m35.373281126s
Apr 21 18:29:25 thanos[22110]: level=info ts=2020-04-21T18:29:25.934654715Z caller=store.go:305 msg="bucket store ready" init_duration=18m17.466030062s
Apr 21 19:23:50 thanos[30304]: level=info ts=2020-04-21T19:23:50.233459402Z caller=store.go:305 msg="bucket store ready" init_duration=15m14.111292242s
Apr 22 23:16:31 thanos[31954]: level=info ts=2020-04-22T23:16:31.374772169Z caller=store.go:305 msg="bucket store ready" init_duration=21m39.163457607s
Apr 23 22:57:16 thanos[29276]: level=info ts=2020-04-23T22:57:16.630899213Z caller=store.go:305 msg="bucket store ready" init_duration=27m54.052617634s
```

Setting compact to delete-delay of 6h reduced our block count enough to cut 7m off of init_duration.
```
Apr 24 18:52:07 thanos[30935]: level=info ts=2020-04-24T18:52:07.461636651Z caller=store.go:305 msg="bucket store ready" init_duration=20m57.610745868s
Apr 24 23:15:36 thanos[30164]: level=info ts=2020-04-24T23:15:36.657906002Z caller=store.go:305 msg="bucket store ready" init_duration=24m37.517365792s
```

The no-meta-json block count is still on the order of our loaded blocks. 
```
thanos_blocks_meta_synced{job="thanos_store",state="no-meta-json"} | 7141
thanos_blocks_meta_synced{job="thanos_store",state="loaded"} | 11388
```

It appears that the deletion mark file is not being removed with the rest of the block, or it is being created after deletion.

**Comment by ThoreKr** [comment]
@gburek-fastly Do these blocks get cleaned up when restarting the compactor?

**Comment by bwplotka** [comment]
Thanks for input! This looks quite insane... cc @pstibrany @pracucci wonder if you can see something immediate, I am busy in other bug fixing for now.

@ThoreKr has good question, maybe it's just our disk cache leaking?

**Comment by pstibrany** [comment]
Sorry, no help yet, just trying to figure this out. If I focus on single block, I'm trying to understand this:

```
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.051849649Z caller=compact.go:271 msg="marking outdated block for deletion" block=01E6J1X5CZT7WFTVMD7P7DD2PS
2020-04-25 12:33:06	level=warn ts=2020-04-25T10:33:06.054120081Z caller=block.go:140 msg="requested to mark for deletion, but file already exists; this should not happen; investigate" err="file 01E6J1X5CZT7WFTVMD7P7DD2PS/deletion-mark.json already exists in bucket"

...
2020-04-25 12:33:06	level=info ts=2020-04-25T10:33:06.056848659Z caller=compact.go:904 msg="start of compactions"

...
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.855760165Z caller=clean.go:58 msg="deleted aborted partial upload" block=01E6J1X5CZT7WFTVMD7P7DD2PS thresholdAge=48h0m0s
2020-04-25 12:33:14	level=info ts=2020-04-25T10:33:14.837906511Z caller=clean.go:48 msg="found partially uploaded block; marking for deletion" block=01E6J1X5CZT7WFTVMD7P7DD2PS
```

First two messages are from "garbage" (duplicate blocks) code from `GarbageCollect` method in compact.go.

After that compactions starts, and after that cleanup of `partial` (no meta.json) blocks is performed. But why does the same block have no `meta.json` file, if previously it was found as a duplicate (so it must have had meta.json file then (?)).

**Comment by pstibrany** [comment]
@bwplotka one bug that I can see is that `DeduplicateFilter` only appends to `duplicateIDs` but never clears it (like at the beginning of `Filter` method), which would explain why we try to garbage collect (by marking as deleted!) same ids again and again.

**Comment by bwplotka** [comment]
Wooooow that's it. (: You got it!

**Comment by bwplotka** [comment]
Fixed on 0.12 `release-0.12` branch, we will release 0.12.2 this week. https://github.com/thanos-io/thanos/pull/2544

**Comment by Xtigyro** [comment]
@bwplotka This is still happening. Using v0.15.
